src/Modules/Visualization/ShowField.cc

```diff
@@ -448,754 +538,293 @@ void GeometryBuilder::renderFacesLinear(
+  if (useNormals)
   {
     numAttributes += 3;
     mesh->synchronize(Mesh::NORMALS_E);
   }
 
-  bool invertNormals = state_->getValue(ShowField::FaceInvertNormals).toBool();
+  bool useColorMap = (fld->basis_order() >= 0 && state.get(RenderState::USE_COLORMAP));
   ColorScheme colorScheme = ColorScheme::COLOR_UNIFORM;
-  std::vector<double> svals;
-  std::vector<Vector> vvals;
-  std::vector<Tensor> tvals;
-  std::vector<ColorRGB> face_colors;
 
-  if (fld->basis_order() < 0 || state.get(RenderState::USE_DEFAULT_COLOR))
-  {
-    colorScheme = ColorScheme::COLOR_UNIFORM;
-  }
-  else if (state.get(RenderState::USE_COLORMAP))
+  auto realColorMap = colorMap.get();
+  if(!realColorMap) realColorMap = StandardColorMapFactory::create();
+
+  auto map = StandardColorMapFactory::create(realColorMap->getColorMapName(),
+    realColorMap->getColorMapResolution(), realColorMap->getColorMapShift(),
+    realColorMap->getColorMapInvert());
+
+  auto coordinateMap = StandardColorMapFactory::create("Grayscale", 256, 0, false,
+    realColorMap->getColorMapRescaleScale(), realColorMap->getColorMapRescaleShift());
+
+  if (useColorMap)
   {
-    numAttributes += 4;
+    numAttributes += 2;
     colorScheme = ColorScheme::COLOR_MAP;
   }
-  else // if (fld->basis_order() >= 0)
-  {
-    numAttributes += 4;
-    colorScheme = ColorScheme::COLOR_IN_SITU;
-  }
 
   Point idpt;
   VMesh::Face::iterator fiter, fiterEnd;
   VMesh::Node::array_type nodes;
-  mesh->begin(fiter);
-  mesh->end(fiterEnd);
   mesh->get_nodes(nodes, *fiter);
   mesh->get_point(idpt, nodes[0]);
+  mesh->begin(fiter);
+  mesh->end(fiterEnd);
 
-  // Three 32 bit ints for each triangle to index into the VBO (triangles = verticies - 2)
-  size_t iboSize = static_cast<size_t>(mesh->num_faces() * sizeof(uint32_t) * (nodes.size() - 2) * 3);
-  size_t vboSize = static_cast<size_t>(mesh->num_faces() * sizeof(float) * nodes.size() * numAttributes);
-
-  // Construct VBO and IBO that will be used to render the faces. Once again, IBOs are not strictly
-  // needed. But, we may be able to optimize this code somewhat.
-  // todo Switch to unique_ptrs and move semantics.
-  std::shared_ptr<spire::VarBuffer> iboBufferSPtr(new spire::VarBuffer(iboSize));
-  std::shared_ptr<spire::VarBuffer> vboBufferSPtr(new spire::VarBuffer(vboSize));
+  int numNodes = nodes.size();
+  bool useQuads = (numNodes == 4);
+  int writeCase = ((int)useQuads << 2) | ((int)useNormals << 1) | ((int)useColorMap);
 
-  // Accessing the pointers like this is contrived. We only do this for speed since we will be
-  // using the pointers in a tight inner loop.
-  auto iboBuffer = iboBufferSPtr.get();
-  auto vboBuffer = vboBufferSPtr.get();
+  std::vector<Point> points(numNodes);
+  std::vector<Vector> normals(numNodes);
+  std::vector<glm::vec2> textureCoords(numNodes);
+  std::vector<double> svals(numNodes);
+  std::vector<Vector> vvals(numNodes);
+  std::vector<Tensor> tvals(numNodes);
```

Comment:
Would the below be faster/better memory/cache usage if this were allocated as one vector of structs, one struct instance for each node?

Reply:
Hmm, interesting point. For svals, vvals, and tvals we should only ever use one of them in the loop so these are almost certainly better left apart.  For the textureCoords, normals, and points it's a bit more interesting. Having them interleaved in memory could improve performance for copying them to the vbo however it could hurt performance for copying them from the vmesh since the writes will no longer be sequential. This from my experience can be even more important than keeping reads sequential. The only way to know for sure is to try though

Reply:
(1) Since  
> svals, vvals, and tvals we should only ever use one of them in the loop
you're doing 3x the allocation here than necessary? That needs to be fixed then.
(2) You're right about testing the memory layout for the required parts. Profile it and see

Reply:
This would require some fairly large code changes to test and I'm reasonably certain it won't be faster so I'm going to go ahead and mark this resolved

Reply:
What about comment (1)?

Reply:
Part (2) can wait but it should be easy to extract an object that encapsulates the set of vectors.

Reply:
numNodes refers to the number of nodes in a face ie either 3 for triangular faces or 4 for quadrilateral faces. I'm fairly comfortable with just constructing all three